### PR TITLE
Locking and cache improvements

### PR DESCRIPTION
**Pull request by adamretter**
This is the culmination of our work at Evolved Binary over the last year on improving the locking and concurrency of eXist-db.

The details are written up in a detailed report and a smaller supporting report:
1. http://www.evolvedbinary.com/technical-reports/exist-db/locking-and-cache-improvements/
2. http://www.evolvedbinary.com/technical-reports/exist-db/asymmetrical-locking/

I will be giving a short introductory presentation on this work at the eXist-db Community Meetup at XML Prague 2018.

### CODE REVIEW

**Comment by adamretter** [comment]
I will rebase this in the morning

**Comment by dizzzz** [comment]
```
    [junit] Running org.exist.storage.lock.DocumentLocksTest
    [junit] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.861 sec
BUILD FAILED
/home/travis/build/eXist-db/exist/build/scripts/junit.xml:153: Test org.exist.storage.lock.DocumentLocksTest failed
```

File: src/org/exist/storage/NativeBroker.java [hunk 1]
```diff
@@ -1917,31 +1917,34 @@ public DocumentImpl storeTempResource(final org.exist.dom.memtree.DocumentImpl d
...
-                Tuple2<Boolean, Collection> tuple = getOrCreateTempCollection(transaction);
-                Collection temp = tuple._2;
-                if (!tuple._1) {
-                    transaction.acquireLock(temp.getLock(), LockMode.WRITE_LOCK);
+                temp = openCollection(XmldbURI.TEMP_COLLECTION_URI, LockMode.WRITE_LOCK);
+
+                // if temp collection does not exist
+                if (temp == null) {
+                    final Tuple2<Boolean, Collection> tuple = getOrCreateTempCollection(transaction);
+                    temp = tuple._2;
```
Comment (dizzzz) [inline src/org/exist/storage/NativeBroker.java]: value can be `null`?

File: src/org/exist/storage/NativeBroker.java [hunk 2]
```diff
@@ -1955,6 +1958,10 @@ public DocumentImpl storeTempResource(final org.exist.dom.memtree.DocumentImpl d
             } catch (final Exception e) {
                 LOG.warn("Failed to store temporary fragment: " + e.getMessage(), e);
                 //abort the transaction
+            } finally {
+                if(openedWithLock && temp != null) {
```
Comment (dizzzz) [inline src/org/exist/storage/NativeBroker.java]: please could you explain? `getOrCreateTempCollection()` does not create a lock on the new collection?

File: extensions/fluent/src/org/exist/fluent/Database.java [hunk 3]
```diff
@@ -604,7 +605,7 @@ public void run() {
 							it.remove();
 						} else {
 							// Must hold write lock on doc before checking stale map to avoid race condition
-							if (doc.getUpdateLock().attempt(LockMode.WRITE_LOCK)) try {
+							try(final ManagedLock<Lock> updateLock = ManagedLock.attempt(doc.getUpdateLock(), LockMode.WRITE_LOCK)) {
```
Comment (dizzzz) [inline extensions/fluent/src/org/exist/fluent/Database.java]: nice!

**Comment by dizzzz** [comment]
impressive amount of work!

**Comment by dizzzz** [comment]
I checked out the branch : 

```
     [echo] --------------------------------------------
     [echo] eXist 4.1.0-SNAPSHOT+201803201413 build
     [echo] --------------------------------------------
     [echo] Git commit: c22a2508b
     [echo] Git branch: locking-and-cache-improvements
```

I see a few (probably related) issues;
- the dashboard shows an empty page
- eXide does not start 

nothing in the logs :-(

**Comment by adamretter** [comment]
@dizzzz Okay so visiting `http://localhost:8080/` redirects you to `http://localhost:8080/exist/apps/dashboard/`. However, instead of the Dashboard, I get a plain collection listing. Which is wrong.

If I look into `$EXIST_HOME/webapp/WEB-INF/logs/urlrewrite.log` then I see the warning:

```
2018-04-02 19:05:55,978 [qtp263382013-29] WARN  (XQueryURLRewrite.java [findSource]:810) - XQueryURLRewrite controller could not be found for path: dashboard/ 
```

I will start debugging this and see what is really happening...

**Comment by adamretter** [comment]
@dizzzz Okay I made a small mistake in my refactoring of `XQueryURLRewrite` in this PR. The code in `XQueryURLRewrite` was very tricky to follow. I have now fixed addressed that in: https://github.com/eXist-db/exist/pull/1795.

So if we can get #1795 merged first, then I will rebase this on that, which should fix your Dashboard issue :-)

**Comment by adamretter** [comment]
@dizzzz Okay that is fixed now :-)

**Comment by dizzzz** [comment]
Failing build for openjdk8 ; restarting

**Comment by dizzzz** [comment]
I am not sure yet, but the money app seems not to be functional. Under investigation

**Comment by duncdrum** [comment]
@dizzzz I have observed that bug where a running query will appear at the top of the window but not in the lower "running queries" section on 4.0.0 (and earlier), so I don't think its specific to this PR.

**Comment by dizzzz** [comment]
My findings...
- the monex UI keeps to be largely disfunctional
- re-indexing took for me twice the original time
- during re-indexing the database is improved available for handling queries (in the past: it was completely blocked)

**Comment by dizzzz** [comment]
I Will check if i can make the PR in sync with current develop and will try ....

**Comment by dizzzz** [comment]
The monex UI is still not working...

**Comment by dizzzz** [comment]
I verified the JMX structure under the hood `http://localhost:8080/exist/status?c=instances&c=processes&c=locking&c=memory&c=caches&c=system&c=operatingsystem&token=XXXXXXXXXX`but the contents look the same, except.... the ordering of elements @wolfgangmm

**Comment by wolfgangmm** [comment]
I just hit a deadlock when trying to update the dashboard. Info attached.

```xml
<jmx:LockTable name="org.exist.management:type=LockTable">
   <jmx:Attempting>
      <jmx:row>
         <jmx:key>/db/apps/dashboard/modules/install.xql</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>DOCUMENT</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:lockMode>WRITE_LOCK</jmx:lockMode>
                     <jmx:ownerThread>qtp1077702221-40</jmx:ownerThread>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
      <jmx:row>
         <jmx:key>/db</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>COLLECTION</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:lockMode>INTENTION_READ</jmx:lockMode>
                     <jmx:ownerThread>qtp1077702221-62</jmx:ownerThread>
                  </jmx:row>
                  <jmx:row>
                     <jmx:lockMode>INTENTION_READ</jmx:lockMode>
                     <jmx:ownerThread>qtp1077702221-55</jmx:ownerThread>
                  </jmx:row>
                  <jmx:row>
                     <jmx:lockMode>INTENTION_READ</jmx:lockMode>
                     <jmx:ownerThread>qtp1077702221-52</jmx:ownerThread>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
   </jmx:Attempting>
   <jmx:Acquired>
      <jmx:row>
         <jmx:key>/db/apps/dashboard/modules/install.xql</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>DOCUMENT</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:key>READ_LOCK</jmx:key>
                     <jmx:value>
                        <jmx:row>
                           <jmx:key>qtp1077702221-40</jmx:key>
                           <jmx:value>1</jmx:value>
                        </jmx:row>
                     </jmx:value>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
      <jmx:row>
         <jmx:key>/db/apps</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>COLLECTION</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:key>WRITE_LOCK</jmx:key>
                     <jmx:value>
                        <jmx:row>
                           <jmx:key>qtp1077702221-40</jmx:key>
                           <jmx:value>1</jmx:value>
                        </jmx:row>
                     </jmx:value>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
      <jmx:row>
         <jmx:key>/db/apps/dashboard/modules</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>COLLECTION</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:key>WRITE_LOCK</jmx:key>
                     <jmx:value>
                        <jmx:row>
                           <jmx:key>qtp1077702221-40</jmx:key>
                           <jmx:value>1</jmx:value>
                        </jmx:row>
                     </jmx:value>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
      <jmx:row>
         <jmx:key>/db/apps/dashboard</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>COLLECTION</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:key>WRITE_LOCK</jmx:key>
                     <jmx:value>
                        <jmx:row>
                           <jmx:key>qtp1077702221-40</jmx:key>
                           <jmx:value>1</jmx:value>
                        </jmx:row>
                     </jmx:value>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
      <jmx:row>
         <jmx:key>/db</jmx:key>
         <jmx:value>
            <jmx:row>
               <jmx:key>COLLECTION</jmx:key>
               <jmx:value>
                  <jmx:row>
                     <jmx:key>WRITE_LOCK</jmx:key>
                     <jmx:value>
                        <jmx:row>
                           <jmx:key>qtp1077702221-40</jmx:key>
                           <jmx:value>1</jmx:value>
                        </jmx:row>
                     </jmx:value>
                  </jmx:row>
               </jmx:value>
            </jmx:row>
         </jmx:value>
      </jmx:row>
   </jmx:Acquired>
</jmx:LockTable>
```

[deadlock.txt](https://github.com/eXist-db/exist/files/1961555/deadlock.txt)

**Comment by adamretter** [comment]
@dizzzz So Monex seems to rely on the structure of the JMX output being exact. If you look at the console in Chrome when you load Monex you will see the error:

```
Uncaught TypeError: Unable to process binding "text: function(){return $data.jmx.LockManager.WaitingThreads().length || 0 }"
Message: Cannot read property 'WaitingThreads' of undefined
    at text (eval at parseBindingsString (knockout-3.1.0.js:56), <anonymous>:3:86)
    at update (knockout-3.1.0.js:76)
    at function.a.h.G (knockout-3.1.0.js:60)
    at h (knockout-3.1.0.js:40)
    at Object.a.ba.a.h (knockout-3.1.0.js:42)
    at knockout-3.1.0.js:60
    at Object.r (knockout-3.1.0.js:8)
    at k (knockout-3.1.0.js:59)
    at h (knockout-3.1.0.js:57)
    at f (knockout-3.1.0.js:57)
```

So the `LockManager` as it was for JMX, is replaced by the new LockTable. Also I updated some stuff for the CollectionCache. However, I guess Monex will need some fixes for the updated JMX output. Also it might be good to modify Monex so that if it doesn't find one element in the XML, it doesn't fail for all the others - and could also provide an error message to the user.

**Comment by dizzzz** [comment]
Right... Hi forgot to investigate this a bit more, but yes, I'd say that monex should be more relaxed here, and more forgiving. Maybe..... t is implemented this way due to performance reasons? @wolfgangmm ?

**Comment by wolfgangmm** [comment]
As for option (3), even if you move all the functionality into Java and expose it as an XQuery function, this function still needs to be called somehow via HTTP. The only solution would be to avoid calling an XQuery altogether, e.g. by hardcoding everything into a servlet. This seems less elegant and much more difficult to maintain.

Concerning option (2), I

**Comment by duncdrum** [comment]
seriously great stuff @adam .
[docs](http://www.exist-db.org/exist/apps/doc/using-collections.xml?field=all&id=D3.15#D3.15)
to quote the author of the evolved binary technical report ;)
```
Determining which pattern is correct is not a simple task, and is currently un-documented in eXist-db. 
```

**Comment by duncdrum** [comment]
see https://github.com/eXist-db/documentation/issues/106

**Pull request merged**